From 58c392c78d49815adfea738fe611db7bac65de75 Mon Sep 17 00:00:00 2001
From: Stephen Yuan Jiang <syuanjiangdev@gmail.com>
Date: Thu, 9 Mar 2017 18:38:57 -0800
Subject: [PATCH 342/362] HBASE-17764 Solve TestMultiSlaveReplication
 flakiness (Stephen Yuan Jiang)

(cherry picked from commit 7dfb3b3194b9096104020f4c87401813c4716dbe)

Change-Id: I972defec4819c3db97ef4393059e0045f4b9b1f5
Author:    Stephen Yuan Jiang <syuanjiangdev@gmail.com>
Reason:    Supportability
Ref:       CDH-53568
---
 .../replication/TestMultiSlaveReplication.java     |   30 ++++++++++++++------
 1 file changed, 22 insertions(+), 8 deletions(-)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestMultiSlaveReplication.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestMultiSlaveReplication.java
index 9a98316..38e5edc 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestMultiSlaveReplication.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestMultiSlaveReplication.java
@@ -49,6 +49,8 @@ import org.apache.hadoop.hbase.testclassification.LargeTests;
 import org.apache.hadoop.hbase.util.Bytes;
 import org.apache.hadoop.hbase.zookeeper.MiniZooKeeperCluster;
 import org.apache.hadoop.hbase.zookeeper.ZooKeeperWatcher;
+import org.junit.After;
+import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
@@ -96,9 +98,6 @@ public class TestMultiSlaveReplication {
         "org.apache.hadoop.hbase.replication.TestMasterReplication$CoprocessorCounter");
 
     utility1 = new HBaseTestingUtility(conf1);
-    utility1.startMiniZKCluster();
-    MiniZooKeeperCluster miniZK = utility1.getZkCluster();
-    new ZooKeeperWatcher(conf1, "cluster1", null, true);
 
     conf2 = new Configuration(conf1);
     conf2.set(HConstants.ZOOKEEPER_ZNODE_PARENT, "/2");
@@ -107,12 +106,8 @@ public class TestMultiSlaveReplication {
     conf3.set(HConstants.ZOOKEEPER_ZNODE_PARENT, "/3");
 
     utility2 = new HBaseTestingUtility(conf2);
-    utility2.setZkCluster(miniZK);
-    new ZooKeeperWatcher(conf2, "cluster2", null, true);
 
     utility3 = new HBaseTestingUtility(conf3);
-    utility3.setZkCluster(miniZK);
-    new ZooKeeperWatcher(conf3, "cluster3", null, true);
 
     table = new HTableDescriptor(tableName);
     HColumnDescriptor fam = new HColumnDescriptor(famName);
@@ -122,9 +117,28 @@ public class TestMultiSlaveReplication {
     table.addFamily(fam);
   }
 
+  @Before
+  public void startup() throws Exception {
+    utility1.startMiniZKCluster();
+    MiniZooKeeperCluster miniZK = utility1.getZkCluster();
+    utility1.setZkCluster(miniZK);
+    new ZooKeeperWatcher(conf1, "cluster1", null, true);
+
+    utility2.setZkCluster(miniZK);
+    new ZooKeeperWatcher(conf2, "cluster2", null, true);
+
+    utility3.setZkCluster(miniZK);
+    new ZooKeeperWatcher(conf3, "cluster3", null, true);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    utility1.shutdownMiniZKCluster();
+  }
+
   @Test(timeout=300000)
   public void testMultiSlaveReplication() throws Exception {
-    LOG.info("testCyclicReplication");
+    LOG.info("Start the testMultiSlaveReplication Test");
     MiniHBaseCluster master = utility1.startMiniCluster();
     utility2.startMiniCluster();
     utility3.startMiniCluster();
-- 
1.7.9.5

